File: src/main/java/pl/yalgrin/playnite/simplesync/service/objects/GameService.java
package pl.yalgrin.playnite.simplesync.service.objects;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.vavr.Tuple;
import io.vavr.Tuple2;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.data.r2dbc.repository.R2dbcRepository;
import org.springframework.http.codec.multipart.FilePart;
import org.springframework.stereotype.Service;
import org.springframework.transaction.reactive.TransactionalOperator;
import pl.yalgrin.playnite.simplesync.config.Constants;
import pl.yalgrin.playnite.simplesync.domain.objects.Game;
import pl.yalgrin.playnite.simplesync.domain.objects.GameDiff;
import pl.yalgrin.playnite.simplesync.dto.ChangeDTO;
import pl.yalgrin.playnite.simplesync.dto.objects.*;
import pl.yalgrin.playnite.simplesync.enums.ObjectType;
import pl.yalgrin.playnite.simplesync.exception.ForceFetchRequiredException;
import pl.yalgrin.playnite.simplesync.exception.ManualSynchronizationRequiredException;
import pl.yalgrin.playnite.simplesync.mapper.objects.GameMapper;
import pl.yalgrin.playnite.simplesync.repository.objects.GameRepository;
import pl.yalgrin.playnite.simplesync.service.ChangeListenerService;
import pl.yalgrin.playnite.simplesync.service.ChangeService;
import pl.yalgrin.playnite.simplesync.service.MetadataService;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.Function;

import static pl.yalgrin.playnite.simplesync.utils.ReactorUtils.toMono;

@Service
@Slf4j
public class GameService extends AbstractObjectWithMetadataService<Game, GameDiff, GameDTO, GameDiffDTO> {

    private final GameRepository gameRepository;
    private final GenreService genreService;
    private final PlatformService platformService;
    private final CompanyService companyService;
    private final CategoryService categoryService;
    private final TagService tagService;
    private final FeatureService featureService;
    private final SeriesService seriesService;
    private final AgeRatingService ageRatingService;
    private final RegionService regionService;
    private final SourceService sourceService;
    private final CompletionStatusService completionStatusService;

    public GameService(GameRepository repository, R2dbcRepository<GameDiff, Long> diffRepository,
                       GameMapper mapper,
                       ChangeService changeService, ChangeListenerService changeListenerService,
                       MetadataService metadataService, ObjectMapper objectMapper,
                       TransactionalOperator transactionalOperator,
                       GenreService genreService, PlatformService platformService, CompanyService companyService,
                       CategoryService categoryService, TagService tagService, FeatureService featureService,
                       SeriesService seriesService, AgeRatingService ageRatingService, RegionService regionService,
                       SourceService sourceService, CompletionStatusService completionStatusService) {
        super(repository, diffRepository, mapper, changeService, changeListenerService, metadataService, objectMapper,
                transactionalOperator);
        this.gameRepository = repository;
        this.genreService = genreService;
        this.platformService = platformService;
        this.companyService = companyService;
        this.categoryService = categoryService;
        this.tagService = tagService;
        this.featureService = featureService;
        this.seriesService = seriesService;
        this.ageRatingService = ageRatingService;
        this.regionService = regionService;
        this.sourceService = sourceService;
        this.completionStatusService = completionStatusService;
    }

    @Override
    public Mono<GameDTO> saveObject(GameDTO dto, String clientId, Flux<FilePart> fileParts, boolean saveFiles) {
        List<ChangeDTO> changes = Collections.synchronizedList(new ArrayList<>());
        return saveGameWithDependencies(dto, clientId, fileParts, saveFiles, changes)
                .as(transactionalOperator::transactional)
                .flatMap(
                        g -> Mono.defer(() -> changeListenerService.publishChanges(changes)).then(Mono.justOrEmpty(g)));
    }

    private Mono<GameDTO> saveGameWithDependencies(GameDTO dto, String clientId,
                                                   Flux<FilePart> fileParts,
                                                   boolean saveFiles, List<ChangeDTO> changes) {
        return saveGenres(dto, clientId, changes)
                .then(savePlatforms(dto, clientId, changes))
                .then(savePublishers(dto, clientId, changes))
                .then(saveDevelopers(dto, clientId, changes))
                .then(saveCategories(dto, clientId, changes))
                .then(saveTags(dto, clientId, changes))
                .then(saveFeatures(dto, clientId, changes))
                .then(saveSeries(dto, clientId, changes))
                .then(saveAgeRatings(dto, clientId, changes))
                .then(saveRegions(dto, clientId, changes))
                .then(saveSource(dto, clientId, changes))
                .then(saveCompletionStatus(dto, clientId, changes))
                .then(saveObjectWithoutPublishing(dto, clientId, fileParts, saveFiles).doOnNext(
                        t -> changes.add(t._2)).map(Tuple2::_1));
    }

    private Mono<List<Tuple2<GenreDTO, ChangeDTO>>> saveGenres(GameDTO dto, String clientId, List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getGenres, GameDTO::setGenres, genreService);
    }

    private Mono<List<Tuple2<PlatformDTO, ChangeDTO>>> savePlatforms(GameDTO dto, String clientId,
                                                                     List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getPlatforms, GameDTO::setPlatforms, platformService);
    }

    private Mono<List<Tuple2<CompanyDTO, ChangeDTO>>> savePublishers(GameDTO dto, String clientId,
                                                                     List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getPublishers, GameDTO::setPublishers, companyService);
    }

    private Mono<List<Tuple2<CompanyDTO, ChangeDTO>>> saveDevelopers(GameDTO dto, String clientId,
                                                                     List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getDevelopers, GameDTO::setDevelopers, companyService);
    }

    private Mono<List<Tuple2<CategoryDTO, ChangeDTO>>> saveCategories(GameDTO dto, String clientId,
                                                                      List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getCategories, GameDTO::setCategories, categoryService);
    }

    private Mono<List<Tuple2<TagDTO, ChangeDTO>>> saveTags(GameDTO dto, String clientId, List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getTags, GameDTO::setTags, tagService);
    }

    private Mono<List<Tuple2<FeatureDTO, ChangeDTO>>> saveFeatures(GameDTO dto, String clientId,
                                                                   List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getFeatures, GameDTO::setFeatures, featureService);
    }

    private Mono<List<Tuple2<SeriesDTO, ChangeDTO>>> saveSeries(GameDTO dto, String clientId, List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getSeries, GameDTO::setSeries, seriesService);
    }

    private Mono<List<Tuple2<AgeRatingDTO, ChangeDTO>>> saveAgeRatings(GameDTO dto, String clientId,
                                                                       List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getAgeRatings, GameDTO::setAgeRatings, ageRatingService);
    }

    private Mono<List<Tuple2<RegionDTO, ChangeDTO>>> saveRegions(GameDTO dto, String clientId,
                                                                 List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDTO::getRegions, GameDTO::setRegions, regionService);
    }

    private Mono<Tuple2<SourceDTO, ChangeDTO>> saveSource(GameDTO dto, String clientId, List<ChangeDTO> changes) {
        return saveObject(dto, clientId, changes, GameDTO::getSource, GameDTO::setSource, sourceService);
    }

    private Mono<Tuple2<CompletionStatusDTO, ChangeDTO>> saveCompletionStatus(GameDTO dto, String clientId,
                                                                              List<ChangeDTO> changes) {
        return saveObject(dto, clientId, changes, GameDTO::getCompletionStatus, GameDTO::setCompletionStatus,
                completionStatusService);
    }

    private <T extends AbstractObjectDTO> Mono<List<Tuple2<T, ChangeDTO>>> saveObjects(GameDTO dto, String clientId,
                                                                                       List<ChangeDTO> changes,
                                                                                       Function<GameDTO, List<T>> dtoGetter,
                                                                                       BiConsumer<GameDTO, List<T>> dtoSetter,
                                                                                       ObjectSaveService<T> service) {
        return Mono.fromSupplier(() -> dtoGetter.apply(dto))
                .flatMapMany(Flux::fromIterable)
                .concatMap(d -> service.saveObjectWithoutPublishing(d, clientId).switchIfEmpty(Mono.justOrEmpty(
                        Tuple.of(d, null)))).collectList()
                .doOnNext(t -> dtoSetter.accept(dto, t.stream().map(Tuple2::_1).toList()))
                .doOnNext(t -> t.stream().map(Tuple2::_2).filter(Objects::nonNull).forEach(changes::add));
    }

    private <T extends AbstractObjectDTO> Mono<Tuple2<T, ChangeDTO>> saveObject(GameDTO dto, String clientId,
                                                                                List<ChangeDTO> changes,
                                                                                Function<GameDTO, T> dtoGetter,
                                                                                BiConsumer<GameDTO, T> dtoSetter,
                                                                                ObjectSaveService<T> service) {
        return Mono.justOrEmpty(dtoGetter.apply(dto))
                .flatMap(d -> service.saveObjectWithoutPublishing(d, clientId)
                        .switchIfEmpty(Mono.justOrEmpty(Tuple.of(d, null))))
                .doOnNext(t -> dtoSetter.accept(dto, t._1))
                .doOnNext(t -> Optional.ofNullable(t._2).ifPresent(changes::add));
    }

    @Override
    public Mono<GameDTO> saveObjectDiff(GameDiffDTO gameDiffDTO, String clientId, Flux<FilePart> fileParts) {
        List<ChangeDTO> changes = Collections.synchronizedList(new ArrayList<>());
        return saveGameWithDependencies(gameDiffDTO, clientId, fileParts, changes).as(
                        transactionalOperator::transactional)
                .flatMap(
                        g -> Mono.defer(() -> changeListenerService.publishChanges(changes)).then(Mono.justOrEmpty(g)));
    }

    private Mono<GameDTO> saveGameWithDependencies(GameDiffDTO dto, String clientId,
                                                   Flux<FilePart> fileParts,
                                                   List<ChangeDTO> changes) {
        return saveGenres(dto, clientId, changes)
                .then(savePlatforms(dto, clientId, changes))
                .then(savePublishers(dto, clientId, changes))
                .then(saveDevelopers(dto, clientId, changes))
                .then(saveCategories(dto, clientId, changes))
                .then(saveTags(dto, clientId, changes))
                .then(saveFeatures(dto, clientId, changes))
                .then(saveSeries(dto, clientId, changes))
                .then(saveAgeRatings(dto, clientId, changes))
                .then(saveRegions(dto, clientId, changes))
                .then(saveSource(dto, clientId, changes))
                .then(saveCompletionStatus(dto, clientId, changes))
                .then(saveObjectDiffWithoutPublishing(dto, clientId, fileParts).doOnNext(
                        t -> changes.add(t._2)).map(Tuple2::_1));
    }

    private <T extends AbstractObjectDTO> Mono<List<Tuple2<T, ChangeDTO>>> saveObjects(GameDiffDTO dto, String clientId,
                                                                                       List<ChangeDTO> changes,
                                                                                       Function<GameDiffDTO, List<T>> dtoGetter,
                                                                                       BiConsumer<GameDiffDTO, List<T>> dtoSetter,
                                                                                       ObjectSaveService<T> service,
                                                                                       String fieldName) {
        if (dto.getChangedFields() == null || !dto.getChangedFields().contains(fieldName)) {
            return Mono.empty();
        }
        return Mono.fromSupplier(() -> dtoGetter.apply(dto))
                .flatMapMany(Flux::fromIterable)
                .concatMap(d -> service.saveObjectWithoutPublishing(d, clientId).switchIfEmpty(Mono.justOrEmpty(
                        Tuple.of(d, null)))).collectList()
                .doOnNext(t -> dtoSetter.accept(dto, t.stream().map(Tuple2::_1).toList()))
                .doOnNext(t -> t.stream().map(Tuple2::_2).filter(Objects::nonNull).forEach(changes::add));
    }

    private <T extends AbstractObjectDTO> Mono<Tuple2<T, ChangeDTO>> saveObject(GameDiffDTO dto, String clientId,
                                                                                List<ChangeDTO> changes,
                                                                                Function<GameDiffDTO, T> dtoGetter,
                                                                                BiConsumer<GameDiffDTO, T> dtoSetter,
                                                                                ObjectSaveService<T> service,
                                                                                String fieldName) {
        if (dto.getChangedFields() == null || !dto.getChangedFields().contains(fieldName)) {
            return Mono.empty();
        }
        return Mono.justOrEmpty(dtoGetter.apply(dto))
                .flatMap(d -> service.saveObjectWithoutPublishing(d, clientId)
                        .switchIfEmpty(Mono.justOrEmpty(Tuple.of(d, null))))
                .doOnNext(t -> dtoSetter.accept(dto, t._1))
                .doOnNext(t -> Optional.ofNullable(t._2).ifPresent(changes::add));
    }

    private Mono<List<Tuple2<GenreDTO, ChangeDTO>>> saveGenres(GameDiffDTO dto, String clientId,
                                                               List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getGenres, GameDiffDTO::setGenres, genreService,
                GameDiffDTO.Fields.GENRES);
    }

    private Mono<List<Tuple2<PlatformDTO, ChangeDTO>>> savePlatforms(GameDiffDTO dto, String clientId,
                                                                     List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getPlatforms, GameDiffDTO::setPlatforms,
                platformService, GameDiffDTO.Fields.PLATFORMS);
    }

    private Mono<List<Tuple2<CompanyDTO, ChangeDTO>>> savePublishers(GameDiffDTO dto, String clientId,
                                                                     List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getPublishers, GameDiffDTO::setPublishers,
                companyService, GameDiffDTO.Fields.PUBLISHERS);
    }

    private Mono<List<Tuple2<CompanyDTO, ChangeDTO>>> saveDevelopers(GameDiffDTO dto, String clientId,
                                                                     List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getDevelopers, GameDiffDTO::setDevelopers,
                companyService, GameDiffDTO.Fields.DEVELOPERS);
    }

    private Mono<List<Tuple2<CategoryDTO, ChangeDTO>>> saveCategories(GameDiffDTO dto, String clientId,
                                                                      List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getCategories, GameDiffDTO::setCategories,
                categoryService, GameDiffDTO.Fields.CATEGORIES);
    }

    private Mono<List<Tuple2<TagDTO, ChangeDTO>>> saveTags(GameDiffDTO dto, String clientId, List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getTags, GameDiffDTO::setTags, tagService,
                GameDiffDTO.Fields.TAGS);
    }

    private Mono<List<Tuple2<FeatureDTO, ChangeDTO>>> saveFeatures(GameDiffDTO dto, String clientId,
                                                                   List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getFeatures, GameDiffDTO::setFeatures, featureService,
                GameDiffDTO.Fields.FEATURES);
    }

    private Mono<List<Tuple2<SeriesDTO, ChangeDTO>>> saveSeries(GameDiffDTO dto, String clientId,
                                                                List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getSeries, GameDiffDTO::setSeries, seriesService,
                GameDiffDTO.Fields.SERIES);
    }

    private Mono<List<Tuple2<AgeRatingDTO, ChangeDTO>>> saveAgeRatings(GameDiffDTO dto, String clientId,
                                                                       List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getAgeRatings, GameDiffDTO::setAgeRatings,
                ageRatingService, GameDiffDTO.Fields.AGE_RATINGS);
    }

    private Mono<List<Tuple2<RegionDTO, ChangeDTO>>> saveRegions(GameDiffDTO dto, String clientId,
                                                                 List<ChangeDTO> changes) {
        return saveObjects(dto, clientId, changes, GameDiffDTO::getRegions, GameDiffDTO::setRegions, regionService,
                GameDiffDTO.Fields.REGIONS);
    }

    private Mono<Tuple2<SourceDTO, ChangeDTO>> saveSource(GameDiffDTO dto, String clientId, List<ChangeDTO> changes) {
        return saveObject(dto, clientId, changes, GameDiffDTO::getSource, GameDiffDTO::setSource, sourceService,
                GameDiffDTO.Fields.SOURCE);
    }

    private Mono<Tuple2<CompletionStatusDTO, ChangeDTO>> saveCompletionStatus(GameDiffDTO dto, String clientId,
                                                                              List<ChangeDTO> changes) {
        return saveObject(dto, clientId, changes, GameDiffDTO::getCompletionStatus, GameDiffDTO::setCompletionStatus,
                completionStatusService, GameDiffDTO.Fields.COMPLETION_STATUS);
    }

    @Override
    protected Mono<Game> findOrCreateEntity(GameDTO dto) {
        return toMono(gameRepository.findByGameIdAndPluginId(dto.getGameId(), dto.getPluginId()))
                .doOnNext(e -> {
                    log.debug("findOrCreateEntity > found entity with id = {} by game id: {} and plugin id: {}",
                            e.getId(),
                            dto.getGameId(), dto.getPluginId());
                    if (!e.isRemoved() && !StringUtils.equals(e.getPlayniteId(), dto.getId())) {
                        throw new ForceFetchRequiredException();
                    }
                })
                .switchIfEmpty(Mono.fromSupplier(() -> createEntityFromDTO(dto))
                        .doOnNext(e -> log.debug("findOrCreateEntity > creating new entity...")));
    }

    @Override
    protected Mono<Game> findOrCreateEntity(GameDiffDTO dto) {
        return toMono(gameRepository.findByGameIdAndPluginId(dto.getGameId(), dto.getPluginId()))
                .doOnNext(e -> {
                    log.debug("findOrCreateEntity > found entity with id = {} by game id: {} and plugin id: {}",
                            e.getId(),
                            dto.getGameId(), dto.getPluginId());
                    if (e.isRemoved()) {
                        throw new ManualSynchronizationRequiredException("Manual synchronization required!");
                    }
                    if (!StringUtils.equals(e.getPlayniteId(), dto.getId())) {
                        throw new ForceFetchRequiredException();
                    }
                })
                .switchIfEmpty(
                        Mono.error(new ManualSynchronizationRequiredException("Manual synchronization required!")));
    }

    @Override
    protected Set<String> getMetadataFields() {
        return Set.of(Constants.ICON, Constants.COVER_IMAGE, Constants.BACKGROUND_IMAGE);
    }

    @Override
    protected void setHex(Game game, String baseName, String md5) {
        if (Constants.ICON.equals(baseName)) {
            game.setIconMd5(md5);
            game.setChanged(true);
        } else if (Constants.COVER_IMAGE.equals(baseName)) {
            game.setCoverImageMd5(md5);
            game.setChanged(true);
        } else if (Constants.BACKGROUND_IMAGE.equals(baseName)) {
            game.setBackgroundImageMd5(md5);
            game.setChanged(true);
        }
    }

    @Override
    protected boolean shouldSaveMetadata(Game game, byte[] bytes, String md5, String basename) {
        if (game.getId() == null) {
            return true;
        }
        String md5ToCompare = null;
        if (Constants.ICON.equals(basename)) {
            md5ToCompare = game.getIconMd5();
        } else if (Constants.COVER_IMAGE.equals(basename)) {
            md5ToCompare = game.getCoverImageMd5();
        } else if (Constants.BACKGROUND_IMAGE.equals(basename)) {
            md5ToCompare = game.getBackgroundImageMd5();
        }
        return md5ToCompare == null || !StringUtils.equals(md5ToCompare, md5);
    }

    @Override
    protected Flux<Game> findObjectToDelete(GameDTO categoryDTO) {
        return gameRepository.findByPlayniteIdAndGameIdAndPluginIdAndRemovedIsFalse(categoryDTO.getId(),
                categoryDTO.getGameId(), categoryDTO.getPluginId());
    }

    @Override
    protected String getMetadataFolder() {
        return Constants.GAME;
    }

    @Override
    protected Game createEntityFromDTO(GameDTO dto) {
        return Game.builder().playniteId(dto.getId()).build();
    }

    @Override
    protected ObjectType getObjectType() {
        return ObjectType.Game;
    }

    @Override
    protected ObjectType getDiffType() {
        return ObjectType.GameDiff;
    }

    @Override
    protected Class<GameDiffDTO> getDiffDtoClass() {
        return GameDiffDTO.class;
    }
}

File: src/main/kotlin/pl/yalgrin/playnite/simplesync/library/service/GameService.kt
package pl.yalgrin.playnite.simplesync.library.service

import org.apache.commons.lang3.Strings
import org.springframework.http.codec.multipart.FilePart
import org.springframework.stereotype.Service
import org.springframework.transaction.ReactiveTransactionManager
import pl.yalgrin.playnite.simplesync.change.dto.ChangeDTO
import pl.yalgrin.playnite.simplesync.change.service.ChangeListenerService
import pl.yalgrin.playnite.simplesync.change.service.ChangeService
import pl.yalgrin.playnite.simplesync.common.config.BACKGROUND_IMAGE
import pl.yalgrin.playnite.simplesync.common.config.COVER_IMAGE
import pl.yalgrin.playnite.simplesync.common.config.GAME
import pl.yalgrin.playnite.simplesync.common.config.ICON
import pl.yalgrin.playnite.simplesync.common.enums.ObjectType
import pl.yalgrin.playnite.simplesync.common.util.first
import pl.yalgrin.playnite.simplesync.common.util.transactional
import pl.yalgrin.playnite.simplesync.exception.ForceFetchRequiredException
import pl.yalgrin.playnite.simplesync.exception.ManualSynchronizationRequiredException
import pl.yalgrin.playnite.simplesync.library.domain.Game
import pl.yalgrin.playnite.simplesync.library.domain.GameDiff
import pl.yalgrin.playnite.simplesync.library.dto.GameDTO
import pl.yalgrin.playnite.simplesync.library.dto.GameDiffDTO
import pl.yalgrin.playnite.simplesync.library.dto.GameDiffFields
import pl.yalgrin.playnite.simplesync.library.dto.LibraryObjectDTO
import pl.yalgrin.playnite.simplesync.library.mapper.GameMapper
import pl.yalgrin.playnite.simplesync.library.repository.GameDiffRepository
import pl.yalgrin.playnite.simplesync.library.repository.GameRepository
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toFlux

@Service
class GameService(
    protected val gameRepository: GameRepository,
    diffRepository: GameDiffRepository,
    mapper: GameMapper,
    changeService: ChangeService,
    changeListenerService: ChangeListenerService,
    metadataService: MetadataService,
    transactionManager: ReactiveTransactionManager,
    protected val genreService: GenreService,
    protected val platformService: PlatformService,
    protected val companyService: CompanyService,
    protected val categoryService: CategoryService,
    protected val tagService: TagService,
    protected val featureService: FeatureService,
    protected val seriesService: SeriesService,
    protected val ageRatingService: AgeRatingService,
    protected val regionService: RegionService,
    protected val sourceService: SourceService,
    protected val completionStatusService: CompletionStatusService
) : BaseLibraryObjectWithDiffService<GameDTO, GameDiffDTO, Game, GameDiff>(
    gameRepository,
    diffRepository,
    mapper,
    changeService,
    changeListenerService,
    metadataService,
    transactionManager
) {

    override fun saveObjectAndPublishChanges(
        dto: GameDTO,
        fileParts: Flux<FilePart>,
        saveFiles: Boolean
    ): Mono<GameDTO> {
        return saveGameWithDependencies(dto, fileParts, saveFiles)
            .transactional(transactionManager)
            .flatMap { t -> changeListenerService.publishChanges(t.generatedChanges).thenReturn(t.savedObject) }
    }

    private fun saveGameWithDependencies(
        dto: GameDTO,
        fileParts: Flux<FilePart>,
        saveFiles: Boolean
    ): Mono<LibrarySaveResult<GameDTO>> {
        return Flux.concat(
            saveRelatedObjects(dto, genreService, { it.genres }, { dto, list -> dto.genres = list }),
            saveRelatedObjects(dto, platformService, { it.platforms }, { dto, list -> dto.platforms = list }),
            saveRelatedObjects(dto, companyService, { it.publishers }, { dto, list -> dto.publishers = list }),
            saveRelatedObjects(dto, companyService, { it.developers }, { dto, list -> dto.developers = list }),
            saveRelatedObjects(dto, categoryService, { it.categories }, { dto, list -> dto.categories = list }),
            saveRelatedObjects(dto, tagService, { it.tags }, { dto, list -> dto.tags = list }),
            saveRelatedObjects(dto, featureService, { it.features }, { dto, list -> dto.features = list }),
            saveRelatedObjects(dto, seriesService, { it.series }, { dto, list -> dto.series = list }),
            saveRelatedObjects(dto, ageRatingService, { it.ageRatings }, { dto, list -> dto.ageRatings = list }),
            saveRelatedObjects(dto, regionService, { it.regions }, { dto, list -> dto.regions = list }),
            saveRelatedObject(dto, sourceService, { it.source }, { dto, r -> dto.source = r }),
            saveRelatedObject(
                dto,
                completionStatusService,
                { it.completionStatus },
                { dto, r -> dto.completionStatus = r })
        ).collectList()
            .map { it.flatten() }
            .flatMap { changes ->
                saveObject(dto, fileParts, saveFiles)
                    .map { LibrarySaveResult(it.savedObject, changes.plus(it.generatedChanges)) }
            }
    }

    private fun <T : LibraryObjectDTO> saveRelatedObjects(
        dto: GameDTO,
        service: LibraryObjectSaveService<T>,
        dtoGetter: (GameDTO) -> List<T>,
        dtoSetter: (GameDTO, List<T>) -> Unit,
    ): Mono<List<ChangeDTO>> {
        return Mono.fromSupplier { dtoGetter(dto) }
            .flatMapMany { Flux.fromIterable(it) }
            .concatMap { d ->
                service.saveObject(d)
                    .switchIfEmpty(Mono.fromSupplier { LibrarySaveResult(d, emptyList()) })
            }
            .collectList()
            .doOnNext { list -> dtoSetter(dto, list.map { it.savedObject }) }
            .flatMapMany { Flux.fromIterable(it) }
            .flatMap { it.generatedChanges.toFlux() }
            .collectList()
    }

    private fun <T : LibraryObjectDTO> saveRelatedObject(
        dto: GameDTO,
        service: LibraryObjectSaveService<T>,
        dtoGetter: (GameDTO) -> T?,
        dtoSetter: (GameDTO, T) -> Unit,
    ): Mono<List<ChangeDTO>> {
        return Mono.fromSupplier { dtoGetter(dto) }
            .flatMap { d ->
                service.saveObject(d)
                    .switchIfEmpty(Mono.fromSupplier { LibrarySaveResult(d, emptyList()) })
            }
            .doOnNext { result -> dtoSetter(dto, result.savedObject) }
            .map { it.generatedChanges }
    }

    override fun findOrCreateEntity(dto: GameDTO): Mono<Game> {
        return Mono.justOrEmpty(dto)
            .filter { d -> d.gameId != null && d.pluginId != null }
            .flatMap { d ->
                gameRepository.findByGameIdAndPluginId(
                    d.gameId!!,
                    d.pluginId!!
                ).first()
            }
            .flatMap { e ->
                log.debug(
                    "findOrCreateEntity > found entity with id = {} by game id: {} and plugin id: {}",
                    e.id,
                    dto.gameId, dto.pluginId
                )
                if (!e.isRemoved && !Strings.CS.equals(e.playniteId, dto.id)) {
                    Mono.error(ForceFetchRequiredException("Force fetch required!"))
                } else {
                    Mono.just(e)
                }
            }
            .switchIfEmpty(
                Mono.fromSupplier { createEntityFromDTO(dto) }
                    .doOnNext { log.debug("findOrCreateEntity > creating new entity...") }
            )
    }

    override fun findOrCreateEntity(dto: GameDiffDTO): Mono<Game> {
        return Mono.justOrEmpty(dto)
            .filter { d -> d.gameId != null && d.pluginId != null }
            .flatMap { gameRepository.findByGameIdAndPluginId(dto.gameId!!, dto.pluginId!!).first() }
            .flatMap { e ->
                log.debug(
                    "findOrCreateEntity > found entity with id = {} by game id: {} and plugin id: {}",
                    e.id,
                    dto.gameId, dto.pluginId
                )
                if (e.isRemoved) {
                    return@flatMap Mono.error(
                        ManualSynchronizationRequiredException("Manual synchronization required!")
                    )
                }
                if (!Strings.CS.equals(e.playniteId, dto.id)) {
                    Mono.error<Game>(ForceFetchRequiredException("Force fetch required!"))
                } else {
                    Mono.just(e)
                }
            }
            .switchIfEmpty(
                Mono.error(ManualSynchronizationRequiredException("Manual synchronization required!"))
            )
    }

    override fun saveObjectDiffAndPublishChanges(
        diffDto: GameDiffDTO,
        fileParts: Flux<FilePart>
    ): Mono<GameDTO> {
        return saveGameDiffWithDependencies(diffDto, fileParts)
            .transactional(transactionManager)
            .flatMap { t -> changeListenerService.publishChanges(t.generatedChanges).thenReturn(t.savedObject) }
    }

    private fun saveGameDiffWithDependencies(
        diffDto: GameDiffDTO,
        fileParts: Flux<FilePart>
    ): Mono<LibrarySaveResult<GameDTO>> = Flux.concat(
        saveRelatedDiffObjects(
            diffDto,
            genreService,
            GameDiffFields.GENRES,
            { it.genres },
            { dto, list -> dto.genres = list }),
        saveRelatedDiffObjects(
            diffDto,
            platformService,
            GameDiffFields.PLATFORMS,
            { it.platforms },
            { dto, list -> dto.platforms = list }),
        saveRelatedDiffObjects(
            diffDto,
            companyService,
            GameDiffFields.PUBLISHERS,
            { it.publishers },
            { dto, list -> dto.publishers = list }),
        saveRelatedDiffObjects(
            diffDto,
            companyService,
            GameDiffFields.DEVELOPERS,
            { it.developers },
            { dto, list -> dto.developers = list }),
        saveRelatedDiffObjects(
            diffDto,
            categoryService,
            GameDiffFields.CATEGORIES,
            { it.categories },
            { dto, list -> dto.categories = list }),
        saveRelatedDiffObjects(diffDto, tagService, GameDiffFields.TAGS, { it.tags }, { dto, list -> dto.tags = list }),
        saveRelatedDiffObjects(
            diffDto,
            featureService,
            GameDiffFields.FEATURES,
            { it.features },
            { dto, list -> dto.features = list }),
        saveRelatedDiffObjects(
            diffDto,
            seriesService,
            GameDiffFields.SERIES,
            { it.series },
            { dto, list -> dto.series = list }),
        saveRelatedDiffObjects(
            diffDto,
            ageRatingService,
            GameDiffFields.AGE_RATINGS,
            { it.ageRatings },
            { dto, list -> dto.ageRatings = list }),
        saveRelatedDiffObjects(
            diffDto,
            regionService,
            GameDiffFields.REGIONS,
            { it.regions },
            { dto, list -> dto.regions = list }),
        saveRelatedDiffObject(
            diffDto,
            sourceService,
            GameDiffFields.SOURCE,
            { it.source },
            { dto, r -> dto.source = r }),
        saveRelatedDiffObject(
            diffDto,
            completionStatusService,
            GameDiffFields.COMPLETION_STATUS,
            { it.completionStatus },
            { dto, r -> dto.completionStatus = r })
    ).collectList()
        .map { it.flatten() }
        .flatMap { changes ->
            saveObjectDiff(diffDto, fileParts)
                .map { LibrarySaveResult(it.savedObject, changes.plus(it.generatedChanges)) }
        }

    private fun <T : LibraryObjectDTO> saveRelatedDiffObjects(
        dto: GameDiffDTO,
        service: LibraryObjectSaveService<T>,
        fieldName: String,
        dtoGetter: (GameDiffDTO) -> List<T>,
        dtoSetter: (GameDiffDTO, List<T>) -> Unit
    ): Mono<List<ChangeDTO>> {
        return Mono.defer {
            if (!dto.changedFields.contains(fieldName)) {
                Mono.empty()
            } else {
                Mono.fromSupplier { dtoGetter(dto) }
                    .flatMapMany { Flux.fromIterable(it) }
                    .concatMap { d ->
                        service.saveObject(d).switchIfEmpty(
                            Mono.fromSupplier {
                                LibrarySaveResult(
                                    savedObject = d,
                                    generatedChanges = emptyList()
                                )
                            }
                        )
                    }.collectList()
                    .doOnNext { t ->
                        dtoSetter(dto, t.map { it.savedObject })
                    }
                    .map { results -> results.flatMap { it.generatedChanges } }
            }
        }
    }

    private fun <T : LibraryObjectDTO> saveRelatedDiffObject(
        dto: GameDiffDTO,
        service: LibraryObjectSaveService<T>,
        fieldName: String,
        dtoGetter: (GameDiffDTO) -> T?,
        dtoSetter: (GameDiffDTO, T) -> Unit
    ): Mono<List<ChangeDTO>> {
        return Mono.defer {
            if (!dto.changedFields.contains(fieldName)) {
                Mono.empty()
            } else {
                Mono.fromSupplier { dtoGetter(dto) }
                    .flatMap { d ->
                        service.saveObject(d).switchIfEmpty(
                            Mono.fromSupplier {
                                LibrarySaveResult(
                                    savedObject = d,
                                    generatedChanges = emptyList()
                                )
                            }
                        )
                    }
                    .doOnNext { t ->
                        dtoSetter(dto, t.savedObject)
                    }
                    .map { results -> results.generatedChanges }
            }
        }
    }

    override fun findObjectToDelete(dto: GameDTO): Flux<Game> {
        return Flux.defer {
            val id = dto.id
            val gameId = dto.gameId
            val pluginId = dto.pluginId
            if (id == null || gameId == null || pluginId == null) {
                Flux.empty()
            } else {
                gameRepository.findByPlayniteIdAndGameIdAndPluginIdAndRemovedIsFalse(id, gameId, pluginId)
            }
        }
    }

    override fun applyFileChange(
        entity: Game,
        baseName: String,
        md5: String?
    ): Mono<Boolean> {
        return Mono.fromSupplier {
            var changed = false
            when (baseName) {
                ICON -> {
                    changed = entity.iconMd5 != md5
                    entity.iconMd5 = md5
                }

                COVER_IMAGE -> {
                    changed = entity.coverImageMd5 != md5
                    entity.coverImageMd5 = md5
                }

                BACKGROUND_IMAGE -> {
                    changed = entity.backgroundImageMd5 != md5
                    entity.backgroundImageMd5 = md5
                }
            }
            entity.isChanged = changed || entity.isChanged
            changed
        }
    }

    override fun getMetadataFolder(): String {
        return GAME
    }

    override fun createEntityFromDTO(dto: GameDTO): Game {
        return Game(
            playniteId = dto.id
        )
    }

    override fun getMetadataFields(): Set<String> {
        return setOf(ICON, COVER_IMAGE, BACKGROUND_IMAGE)
    }

    override fun hasFileDataChanged(
        entity: Game,
        bytes: ByteArray,
        md5: String,
        fieldName: String
    ): Mono<Boolean> {
        return Mono.fromSupplier {
            if (entity.id == null) {
                return@fromSupplier true
            }
            var md5ToCompare: String? = null
            when (fieldName) {
                ICON -> {
                    md5ToCompare = entity.iconMd5
                }

                COVER_IMAGE -> {
                    md5ToCompare = entity.coverImageMd5
                }

                BACKGROUND_IMAGE -> {
                    md5ToCompare = entity.backgroundImageMd5
                }
            }
            md5ToCompare == null || !Strings.CS.equals(md5ToCompare, md5)
        }
    }

    override fun getObjectType(): ObjectType = ObjectType.GAME

    override fun getDiffType(): ObjectType = ObjectType.GAME_DIFF
}